src/cbmc/bmc.cpp

```diff
@@ -596,3 +599,93 @@ void bmct::setup_unwind()
   if(options.get_option("unwind")!="")
     symex.set_unwind_limit(options.get_unsigned_int_option("unwind"));
 }
+
+int bmct::do_language_agnostic_bmc(
+    const optionst &opts,
+    const goto_modelt &goto_model,
+    const ui_message_handlert::uit &ui,
+    messaget *const message)
+{
+  cbmc_solverst solvers(
+      opts,
+      goto_model.symbol_table,
+      message->get_message_handler());
+  solvers.set_ui(ui);
+
+  message_handlert &mh=message->get_message_handler();
+  safety_checkert::resultt result;
+  goto_symext::branch_worklistt worklist;
+  try
+  {
+    {
+      std::unique_ptr<cbmc_solverst::solvert> cbmc_solver;
+      cbmc_solver=solvers.get_solver();
+      prop_convt &pc=cbmc_solver->prop_conv();
+      bmct bmc(opts, goto_model.symbol_table, mh, pc, worklist);
+      bmc.set_ui(ui);
+      result=bmc.run(goto_model.goto_functions);
+    }
+    INVARIANT(opts.get_bool_option("paths") || worklist.empty(),
+        "Some branches were unexplored when model-checking "
+        "the entire program.");
+
+    while(!worklist.empty())
+    {
+      message->status() << "___________________________\n"
+        << "Starting new path (" << worklist.size() << " to go)\n"
+        << message->eom;
+      std::unique_ptr<cbmc_solverst::solvert> cbmc_solver;
+      cbmc_solver=solvers.get_solver();
+      prop_convt &pc=cbmc_solver->prop_conv();
+      goto_symext::branch_pointt &resume=worklist.front();
+      path_explorert pe(opts,
+          goto_model.symbol_table, mh, pc, resume.equation,
+          resume.state, worklist);
+      result&=pe.run(goto_model.goto_functions);
+      worklist.pop_front();
+    }
+  }
+  catch(const char *error_msg)
```

Comment:
Recommend throwing structured errors rather than plain strings wherever possible

Reply:
@smowton this catch is to deal with a throw from `cbmc_solverst`, i.e. it's code that predates this patchset. As far as I can see, that code only ever throws the number `0` (and it prints a message to `messaget::error()` just before it throws. I'm not familiar with cprover's error handling, what sort of error would you recommend that I throw from `cbmc_solvert` instead of `0`? Or should I just leave that there?

Reply:
"Pretty crap," in brief ;) If it's old code you're catching from, I guess you intend to catch everything? In that case I'd catch `const char *` and `const std::string &` (perhaps mergable if exception catching permits implicit construction to catch, but I doubt it) and then `catch(...)` to catch other errors. If you specifically intend to catch *some* things but let others get by then you should at least document your intent in comments, and probably change the throwing code to throw something more specific.

Reply:
#include "martin_comment_about_catch_all_vs_invariants.h"

If you are going to catch (...) please exit and / or throw upwards; many exceptions are non-recoverable.